vnext/Microsoft.ReactNative.Cxx/JSI/JsiApi.h

```diff
@@ -7,8 +7,55 @@
 #include "../ReactContext.h"
 #include "JsiAbiApi.h"
 
+// Use __ImageBase to get current DLL handle.
+// http://blogs.msdn.com/oldnewthing/archive/2004/10/25/247180.aspx
+extern "C" IMAGE_DOS_HEADER __ImageBase;
+
 namespace winrt::Microsoft::ReactNative {
 
+// Get JSI Runtime from the current JS dispatcher thread.
+// If it is not found, then create it and store it in the context.Properties().
+// Make sure that the JSI runtime holder is removed when the instance is unloaded.
+facebook::jsi::Runtime &GetOrCreateContextRuntime(ReactContext const &context) noexcept {
+  ReactDispatcher jsDispatcher = context.JSDispatcher();
+  VerifyElseCrashSz(jsDispatcher.HasThreadAccess(), "Must be in JS thread");
+
+  // The JSI runtime is not available if we do Web debugging when JS is running in web browser.
+  JsiRuntime abiJsiRuntime = context.Handle().JSRuntime().as<JsiRuntime>();
+  VerifyElseCrashSz(abiJsiRuntime, "JSI runtime is not available");
+
+  // See if the JSI runtime was previously created.
+  JsiAbiRuntime *runtime = JsiAbiRuntime::GetFromJsiRuntime(abiJsiRuntime);
+  if (!runtime) {
+    // Create the runtime
+    std::unique_ptr<JsiAbiRuntime> runtimeHolder = std::make_unique<JsiAbiRuntime>(abiJsiRuntime);
+    runtime = runtimeHolder.get();
+
+    // We want to keep the JSI runtime while current instance is alive.
+    // The JSI runtime object must be local to our DLL.
+    // We create a property name based on the current DLL handle.
+    HMODULE currentDll = reinterpret_cast<HMODULE>(&__ImageBase);
```

Comment:
isn't this tying the JSI to the m.rn dll? it should be possible to have more than one JS instance running in the same process (against the same m.rn dll)

Reply:
No, it does not prevent running multiple instances in the same process. We are creating a property in a specific React instance. Different instances can have property with the same name, but different values. But they must use different threads. We keep the non-abi-safe data per DLL in order to make sure that different DLLs do not use non-abi-safe data from each other.

---
In reply to: [542910378](https://github.com/microsoft/react-native-windows/pull/6745#discussion_r542910378) [](ancestors = 542910378)

Reply:
@vmoroz is this documented somewhere? if not, can you please add it to the API documentation?

Reply:
I plan to do the doc updates next week. This week I want to make sure that the core functionality works.

---
In reply to: [543014346](https://github.com/microsoft/react-native-windows/pull/6745#discussion_r543014346) [](ancestors = 543014346)

Reply:
Filed documentation IOU: #6748